# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: modules/libjar/nsZipArchive.h
# Commit: 4208ed5fc3b0
# Full Hash: 4208ed5fc3b02fddc3624805ff3937a68a4aa4f6
# Author: Doug Thayer <dothayer@mozilla.com>
# Date: 2020-07-08 09:42:17
# Regressor Bug: 1627075
# File Overlap Count: 1
# Description:
#   Bug 1627075 - Build Omnijar file list from startup cache r=froydnj
#   
#   We would like to be able to defer opening the omnijar files until after startup
#   if the StartupCache has already been populated. Opening the omnijar files takes
#   a nontrivial time, at least on Windows, and almost everything in the omnijar
# ==============================================================================

diff -r 14420a624d0a -r 4208ed5fc3b0 modules/libjar/nsZipArchive.h
--- a/modules/libjar/nsZipArchive.h	Tue Jul 07 15:19:09 2020 +0000
+++ b/modules/libjar/nsZipArchive.h	Wed Jul 08 02:43:02 2020 +0000
@@ -97,11 +97,17 @@
    * object. If we were allowed to use exceptions this would have been
    * part of the constructor
    *
-   * @param   aZipHandle  The nsZipHandle used to access the zip
-   * @param   aFd         Optional PRFileDesc for Windows readahead optimization
+   * @param   aZipHandle      The nsZipHandle used to access the zip
+   * @param   aFd             Optional PRFileDesc for Windows readahead
+                              optimization
+   * @param   aCachedCentral  Optional cached buffer containing the zip central
+                              for this zip.
+   * @param   aCachedCentralSize  Optional size of aCachedCentral.
    * @return  status code
    */
-  nsresult OpenArchive(nsZipHandle* aZipHandle, PRFileDesc* aFd = nullptr);
+  nsresult OpenArchive(nsZipHandle* aZipHandle, PRFileDesc* aFd = nullptr,
+                       const uint8_t* aCachedCentral = nullptr,
+                       size_t aCachedCentralSize = 0);
 
   /**
    * OpenArchive
@@ -109,9 +115,13 @@
    * Convenience function that generates nsZipHandle
    *
    * @param   aFile         The file used to access the zip
+   * @param   aCachedCentral  Optional cached buffer containing the zip central
+                              for this zip.
+   * @param   aCachedCentralSize  Optional size of aCachedCentral.
    * @return  status code
    */
-  nsresult OpenArchive(nsIFile* aFile);
+  nsresult OpenArchive(nsIFile* aFile, const uint8_t* aCachedCentral = nullptr,
+                       size_t aCachedCentralSize = 0);
 
   /**
    * Test the integrity of items in this archive by running
@@ -181,6 +191,17 @@
    */
   const uint8_t* GetData(nsZipItem* aItem);
 
+  /**
+   * Copies the contents of the zip central directory, and returns it to the
+   * caller to take ownership. This is useful for caching the contents of the
+   * central directory, which can be compressed and stored elsewhere, and
+   * passed back into OpenArchive when this archive is opened in the future.
+   *
+   * @param   aSize       size_t pointer to be filled with the size of the
+                          returned buffer.
+   */
+  mozilla::UniquePtr<uint8_t[]> CopyCentralDirectoryBuffer(size_t* aSize);
+
   bool GetComment(nsACString& aComment);
 
   /**
@@ -204,6 +225,8 @@
   mozilla::ArenaAllocator<1024, sizeof(void*)> mArena;
 
   const char* mCommentPtr;
+  size_t mZipCentralOffset;
+  size_t mZipCentralSize;
   uint16_t mCommentLen;
 
   // Whether we synthesized the directory entries
@@ -223,6 +246,7 @@
   //--- private methods ---
   nsZipItem* CreateZipItem();
   nsresult BuildFileList(PRFileDesc* aFd = nullptr);
+  nsresult BuildFileListFromBuffer(const uint8_t* aBuf, const uint8_t* aEnd);
   nsresult BuildSynthetics();
 
   nsZipArchive& operator=(const nsZipArchive& rhs) = delete;